### PR TITLE
Add logging request data to MDC for display in all log messages

### PR DESCRIPTION
**Pull request by mapingo**

### CODE REVIEW

File: generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.ext.Provider;
+
+import org.apache.commons.io.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.MDC;
+
+@Provider
+public class LoggerRequestDataFilter implements ContainerRequestFilter, ContainerResponseFilter {
```
Comment (mapingo) [inline generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java]: Java docs?

File: generators/messaging-adapter-core/src/main/java/uk/gov/justice/services/adapter/messaging/JmsLoggerMetadataInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import javax.inject.Inject;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.InvocationContext;
+import javax.jms.TextMessage;
+import javax.json.JsonObjectBuilder;
+
+import org.slf4j.Logger;
+import org.slf4j.MDC;
+
+public class JmsLoggerMetadataInterceptor {
```
Comment (mapingo) [inline generators/messaging-adapter-core/src/main/java/uk/gov/justice/services/adapter/messaging/JmsLoggerMetadataInterceptor.java]: Java docs

File: common/src/main/java/uk/gov/justice/services/common/log/LoggerConstants.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.justice.services.common.log;
+
+public final class LoggerConstants {
+
+    public static final String REQUEST_DATA = "requestData";
+    public static final String SERVICE_CONTEXT = "serviceContext";
+    public static final String METADATA = "metadata";
+
+    private LoggerConstants() {
```
Comment (matt-rich) [inline common/src/main/java/uk/gov/justice/services/common/log/LoggerConstants.java]: Could you make this abstract and remove the default private constructor?

File: generators/messaging-adapter-core/src/main/java/uk/gov/justice/services/adapter/messaging/JmsParametersChecker.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package uk.gov.justice.services.adapter.messaging;
+
+import static java.lang.String.format;
+
+import javax.jms.TextMessage;
+
+/**
+ * Checks the parameters received by a JMS Interceptor
+ */
+public class JmsParametersChecker {
```
Comment (matt-rich) [inline generators/messaging-adapter-core/src/main/java/uk/gov/justice/services/adapter/messaging/JmsParametersChecker.java]: So we don't have happy path tests due to the static MDC?

File: generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+/**
+ * Filter listens to all requests and adds request information to the Logger Mapped Diagnostic
+ * Context.  This can be added to the log output by setting %X{requestData} in the logger pattern.
+ */
+@Provider
+public class LoggerRequestDataFilter implements ContainerRequestFilter, ContainerResponseFilter {
+
+    private static final String JSON_MEDIA_TYPE_FORMAT = "+json";
+    private static final String PAYLOAD_METADATA = "_metadata";
```
Comment (matt-rich) [inline generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java]: Should we reuse these constants from elsewhere?

File: generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+                if (!userId.isPresent()) {
+                    userId = metadata.userId();
+                }
+            }
+
+            requestContext.setEntityStream(new ByteArrayInputStream(payload.getBytes(charset)));
+        }
+
+        if (id.isPresent() || name.isPresent() || correlationId.isPresent() || sessionId.isPresent() || userId.isPresent()) {
```
Comment (matt-rich) [inline generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java]: Do you need the sessionId etc up here too?

File: common/src/main/java/uk/gov/justice/services/common/log/LoggerConstants.java (hunk 3 above)
Comment (mapingo) [inline common/src/main/java/uk/gov/justice/services/common/log/LoggerConstants.java]: You are correct final abstract is not allowed.  This seems to be the best we can do for reusable constants across multiple classes

File: generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        mergeHeadersWithPayloadMetadataIfPresent(requestContext, headers)
+                .ifPresent(metadataBuilder -> builder.add(METADATA, metadataBuilder));
+
+        MDC.put(REQUEST_DATA, builder.build().toString());
+
+        trace(logger, () -> "Request data added to MDC");
+    }
+
+    @Override
+    public void filter(final ContainerRequestContext requestContext, final ContainerResponseContext responseContext) throws IOException {
```
Comment (mapingo) [inline generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java]: I have extracted common code, but due to uncommon usage I have avoided abstract class

File: generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java (hunk 6 above)
Comment (mapingo) [inline generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java]: The reason for this was to remove completely the metadata object if it was not required.

File: generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java [hunk 8]
```diff
@@ -145,6 +114,62 @@ private void addContentTypeAndAcceptIfPresent(final JsonObjectBuilder builder, f
         return Optional.empty();
     }
 
+    private Optional<String> idFromHeaderOrMetadataIfPresent(final MultivaluedMap<String, String> headers, final Optional<Metadata> payloadMetadata) {
+        return Optional.ofNullable(headers.getFirst(ID))
```
Comment (matt-rich) [inline generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java]: could we pass this to a private method that takes a NAME and a lambda of the metadata -> metadata.id().toString() bit?

Would that look cleaner?

File: generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java [hunk 9]
```diff
@@ -145,6 +114,62 @@ private void addContentTypeAndAcceptIfPresent(final JsonObjectBuilder builder, f
...
+    private Optional<Metadata> payloadMetadataIfPresent(final ContainerRequestContext requestContext) throws IOException {
+        final Optional<MediaType> mediaType = Optional.ofNullable(requestContext.getMediaType());
+
+        if (isPayloadPresent(mediaType)) {
+            final String charset = charsetFrom(mediaType.get());
+            final String payload = IOUtils.toString(requestContext.getEntityStream(), charset);
+
+            try {
+
+                if (payload.contains(JsonEnvelope.METADATA)) {
```
Comment (matt-rich) [inline generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java]: static import?

File: generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java (hunk 8 above)
Comment (matt-rich) [inline generators/rest-adapter-core/src/main/java/uk/gov/justice/services/adapter/rest/filter/LoggerRequestDataFilter.java]: Never mind you did this

**Pull request merged**